ssl/quic/qlog_event_helpers.c

```diff
@@ -181,3 +182,432 @@ void ossl_qlog_event_recovery_packet_lost(QLOG *qlog,
+                goto unknown;
+
+            QLOG_STR("frame_type", "stream_data_blocked");
+            QLOG_U64("stream_id", stream_id);
+            QLOG_U64("limit", x);
+        }
+        break;
+    case OSSL_QUIC_FRAME_TYPE_STREAMS_BLOCKED_BIDI:
+    case OSSL_QUIC_FRAME_TYPE_STREAMS_BLOCKED_UNI:
+        {
+            uint64_t x;
+
+            if (!ossl_quic_wire_decode_frame_streams_blocked(pkt, &x))
+                goto unknown;
+
+            QLOG_STR("frame_type", "streams_blocked");
+            QLOG_STR("stream_type",
+                     frame_type == OSSL_QUIC_FRAME_TYPE_STREAMS_BLOCKED_BIDI
+                        ? "bidirectional" : "unidirectional");
+            QLOG_U64("limit", x);
+        }
+        break;
+    case OSSL_QUIC_FRAME_TYPE_NEW_TOKEN:
+        {
+            const unsigned char *token;
+            size_t token_len;
+
+            if (!ossl_quic_wire_decode_frame_new_token(pkt, &token, &token_len))
+                goto unknown;
+
+            QLOG_STR("frame_type", "new_token");
+            QLOG_BEGIN("token");
+                QLOG_BEGIN("raw");
+                    QLOG_BIN("data", token, token_len);
+                QLOG_END();
+            QLOG_END();
+        }
+        break;
+    default:
+unknown:
+        QLOG_STR("frame_type", "unknown");
+        QLOG_U64("frame_type_value", frame_type);
+
+        /*
+         * Can't continue scanning for frames in this case as the frame length
+         * is unknown. We log the entire body the rest of the packet payload
+         * as the raw data of the frame.
+         */
+        QLOG_BEGIN("raw");
+            QLOG_BIN("data", PACKET_data(&orig_pkt),
+                     PACKET_remaining(&orig_pkt));
+        QLOG_END();
+        ignore_res(PACKET_forward(pkt, PACKET_remaining(pkt)));
+        break;
+    }
+
+    return 1;
+}
+
+static int log_frames(QLOG *qlog_instance,
+                      const OSSL_QTX_IOVEC *iovec,
+                      size_t num_iovec)
+{
+    size_t i;
+    PACKET pkt;
+    size_t rem_before, rem_after;
+    uint64_t need_skip = 0;
+
+    for (i = 0; i < num_iovec; ++i) {
+        if (!PACKET_buf_init(&pkt, iovec[i].buf, iovec[i].buf_len))
+            return 0;
+
+        while (PACKET_remaining(&pkt) > 0) {
+            if (need_skip > 0) {
+                size_t adv = need_skip;
+
+                if (adv < PACKET_remaining(&pkt))
+                    adv = PACKET_remaining(&pkt);
+
+                if (!PACKET_forward(&pkt, adv))
+                    return 0;
+
+                need_skip -= adv;
+                continue;
+            }
+
+            rem_before = PACKET_remaining(&pkt);
+
+            QLOG_BEGIN(NULL)
+            {
+                if (!log_frame(qlog_instance, &pkt, &need_skip))
+                    goto cont;
+
+                rem_after = PACKET_remaining(&pkt);
+                QLOG_U64("length", rem_before - rem_after);
```

Comment:
Isn't the length already encoded as part of log_frame above?

Reply:
Looks like an errant line pertaining to PING only. Fixed.